File: src/engines/PolicyEngine.ts
/**
 * PolicyEngine — Google Play Compliance Gatekeeper
 * أول طبقة دفاع: تفحص أي طلب قبل أي عملية
 * لا يمر أي كود أو اقتراح بدون موافقتها
 */

export type PolicyVerdict = 'allowed' | 'blocked' | 'warning';

export interface PolicyResult {
  verdict:     PolicyVerdict;
  violations:  PolicyViolation[];
  riskScore:   number;   // 0-100
  summary:     string;
}

export interface PolicyViolation {
  rule:        string;
  description: string;
  severity:    'critical' | 'high' | 'medium';
  evidence:    string;
}

// ===== Policy Rules =====
interface PolicyRule {
  id:          string;
  description: string;
  severity:    'critical' | 'high' | 'medium';
  test:        (code: string) => string | null;  // returns evidence string or null if clean
}

const POLICY_RULES: PolicyRule[] = [

  // ── CRITICAL: Malware Patterns ──────────────────────────────────
  {
    id: 'POL001', severity: 'critical',
    description: 'كود يحذف ملفات النظام أو يمسح البيانات',
    test: code => {
      const patterns = [
        /(?:shutil\.rmtree|os\.remove|os\.unlink)\s*\(\s*['"]\/(?:etc|sys|bin|boot|usr|root)/i,
        /subprocess.*(?:rm\s+-rf\s+\/|del\s+\/[sqf]+\s+[A-Z]:\\)/i,
        /format\s*(?:c:|\/dev\/)/i,
        /os\.system\s*\(\s*['"](?:rm|del|format)/i,
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  {
    id: 'POL002', severity: 'critical',
    description: 'محاولة سرقة بيانات المستخدم أو كلمات المرور',
    test: code => {
      const patterns = [
        /(?:keylog|keystroke|screenshot.*send|capture.*password)/i,
        /(?:getpass|input).*(?:send|upload|post|requests\.)/i,
        /open\s*\(\s*['"](?:\/etc\/(?:passwd|shadow)|.*\.(?:kdbx|keychain))/i,
        /(?:steal|harvest|exfiltrate|leak).*(?:password|credential|token|key)/i,
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  {
    id: 'POL003', severity: 'critical',
    description: 'برمجية خبيثة: ransomware أو تشفير ملفات المستخدم',
    test: code => {
      const patterns = [
        /(?:encrypt|AES|Fernet).*(?:walk|glob|rglob).*\.(?:doc|pdf|jpg|png|txt)/i,
        /for.*(?:file|path).*(?:encrypt|cipher|lock)/i,
        /ransom|cryptolocker|wannacry/i,
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  {
    id: 'POL004', severity: 'critical',
    description: 'أدوات اختراق أو هجمات: port scanner, brute force, exploit',
    test: code => {
      const patterns = [
        /(?:brute.?force|password.?spray|credential.?stuff)/i,
        /for.*port.*socket\.connect.*except/i,
        /(?:reverse.?shell|bind.?shell|nc\s+-e|netcat.*-e)/i,
        /(?:metasploit|meterpreter|cobalt.?strike|mimikatz)/i,
        /(?:sql.?inject|xss.?payload|buffer.?overflow)/i,
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  {
    id: 'POL005', severity: 'critical',
    description: 'إخفاء كود (Obfuscation): base64 + exec أو eval مشفر',
    test: code => {
      const patterns = [
        /exec\s*\(\s*(?:base64|b64|decode|decompress)/i,
        /eval\s*\(\s*(?:base64|b64|__import__|compile)/i,
        /(?:base64\.b64decode|binascii\.unhexlify).*exec/i,
        /(?:zlib\.decompress|gzip\.decompress).*exec/i,
        /chr\(\d+\)\s*\+\s*chr\(\d+\)/,   // chr(x)+chr(y) obfuscation
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  {
    id: 'POL006', severity: 'critical',
    description: 'تجاوز صلاحيات النظام أو privilege escalation',
    test: code => {
      const patterns = [
        /(?:setuid|setgid|chmod\s*(?:777|4755)|chown\s+root)/i,
        /(?:sudo|runas|elevate).*(?:silent|quiet|-y|-yes)/i,
        /ctypes.*(?:windll|kernel32|advapi32).*(?:privilege|token|admin)/i,
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  // ── HIGH: Suspicious Patterns ────────────────────────────────────
  {
    id: 'POL007', severity: 'high',
    description: 'تحميل وتشغيل ملفات من الإنترنت مباشرةً',
    test: code => {
      const patterns = [
        /(?:urllib|requests|wget).*(?:download|get).*exec/i,
        /(?:download|fetch).*(?:\.exe|\.sh|\.bat|\.ps1).*(?:run|execute|system)/i,
        /requests\.get.*\.content.*open.*\.exe/i,
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  {
    id: 'POL008', severity: 'high',
    description: 'إرسال بيانات خارجية بشكل مخفي (data exfiltration)',
    test: code => {
      const patterns = [
        /requests\.(post|put).*(?:password|key|secret|token|credential)/i,
        /socket\.connect.*(?:send|sendall).*(?:os\.environ|getpass|input)/i,
        /(?:smtp|ftp|curl).*(?:secret|password|credential)/i,
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  {
    id: 'POL009', severity: 'high',
    description: 'تعديل ملفات النظام أو السجل (Registry)',
    test: code => {
      const patterns = [
        /winreg.*(?:HKEY_LOCAL_MACHINE|HKLM).*(?:SetValue|CreateKey)/i,
        /open\s*\(\s*['"]\/etc\/(?:hosts|crontab|sudoers|rc\.)/i,
        /(?:crontab|at\s+|schtasks).*(?:add|create|modify)/i,
      ];
      for (const p of patterns) {
        const m = code.match(p);
        if (m) return m[0].substring(0, 80);
      }
      return null;
    },
  },

  {
    id: 'POL010', severity: 'high',
    description: 'أوامر نظام خطيرة بدون context واضح',
    test: code => {
      const dangerous = ['dd if=', 'mkfs.', ':(){:|:&};:', '> /dev/sd', 'fork bomb'];
      for (const d of dangerous) {
        if (code.toLowerCase().includes(d.toLowerCase()))
          return `Found: ${d}`;
      }
      return null;
    },
  },

  // ── MEDIUM: Best Practice Violations ────────────────────────────
  {
    id: 'POL011', severity: 'medium',
    description: 'اتصال شبكي بدون timeout أو error handling',
    test: code => {
      if (/requests\.(get|post|put|delete)\s*\(/.test(code) &&
          !/timeout\s*=/.test(code)) {
        return 'requests call without timeout';
      }
      return null;
    },
  },

  {
    id: 'POL012', severity: 'medium',
    description: 'استخدام __import__ أو importlib ديناميكياً — قد يكون خطراً',
    test: code => {
      const m = code.match(/__import__\s*\(|importlib\.import_module\s*\([^'"]/);
      return m ? m[0].substring(0, 60) : null;
    },
  },
];

// ===== Main Policy Check =====
export function checkPolicy(code: string): PolicyResult {
  const violations: PolicyViolation[] = [];

  for (const rule of POLICY_RULES) {
    const evidence = rule.test(code);
    if (evidence) {
      violations.push({
        rule:        rule.id,
        description: rule.description,
        severity:    rule.severity,
        evidence,
      });
    }
  }

  // Calculate risk score
  let riskScore = 0;
  for (const v of violations) {
    if (v.severity === 'critical') riskScore += 40;
    else if (v.severity === 'high') riskScore += 20;
    else riskScore += 8;
  }
  riskScore = Math.min(100, riskScore);

  // Determine verdict
  const hasCritical = violations.some(v => v.severity === 'critical');
  const hasHigh     = violations.some(v => v.severity === 'high');

  let verdict: PolicyVerdict = 'allowed';
  if (hasCritical) verdict = 'blocked';
  else if (hasHigh || violations.length >= 2) verdict = 'warning';

  const summary = verdict === 'blocked'
    ? `مرفوض: ${violations.filter(v => v.severity === 'critical').length} انتهاك حرج`
    : verdict === 'warning'
    ? `تحذير: ${violations.length} مشكلة تحتاج مراجعة`
    : 'نظيف: لا انتهاكات للسياسة';

  return { verdict, violations, riskScore, summary };
}

File: android/app/src/main/java/com/naif/realityengine/PolicyEngine.java
package com.naif.realityengine;

import java.util.*;
import java.util.regex.*;

/**
 * PolicyEngine v3.0
 * ✅ Policy check       — check()
 * ✅ Security rules     — 11 rules (POL001–POL011)
 * ✅ Dangerous APIs     — stripCommentsAndStrings()
 * ✅ Allow/Block        — PolicyVerdict
 * ✅ Risk level         — riskScore 0–100
 * ✅ Confidence score   — confidence() per violation
 * ✅ Explain            — explain() تفسير بالعربي
 * ✅ Whitelist          — addWhitelist() / isWhitelisted()
 * ✅ Custom rules       — addRule()
 * ✅ Batch check        — checkAll()
 */
public class PolicyEngine {

    // ═══════════════════════════════════════════════════
    // Enums & Data Classes
    // ═══════════════════════════════════════════════════

    public enum PolicyVerdict { ALLOWED, WARNING, BLOCKED }

    public static class PolicyViolation {
        public final String  rule;
        public final String  description;
        public final String  evidence;
        public final int     line;
        public final boolean critical;
        public final double  confidence; // 0.0 – 1.0

        public PolicyViolation(String rule, String description,
                                String evidence, int line,
                                boolean critical, double confidence) {
            this.rule        = rule;
            this.description = description;
            this.evidence    = evidence;
            this.line        = line;
            this.critical    = critical;
            this.confidence  = confidence;
        }

        @Override public String toString() {
            return String.format("[%s] %s (س%d) conf=%.0f%% %s",
                rule, description, line,
                confidence * 100,
                critical ? "🔴" : "🟡");
        }
    }

    public static class PolicyResult {
        public final PolicyVerdict          verdict;
        public final List<PolicyViolation>  violations;
        public final int                    riskScore;
        public final String                 summary;
        public final double                 overallConfidence; // متوسط confidence الانتهاكات

        public PolicyResult(PolicyVerdict verdict,
                            List<PolicyViolation> violations,
                            int riskScore) {
            this.verdict           = verdict;
            this.violations        = violations;
            this.riskScore         = riskScore;
            this.summary           = buildSummary(verdict, violations);
            this.overallConfidence = violations.isEmpty() ? 1.0
                : violations.stream().mapToDouble(v -> v.confidence).average().orElse(0.0);
        }

        private static String buildSummary(PolicyVerdict verdict,
                                            List<PolicyViolation> violations) {
            if (violations.isEmpty()) return "✅ نظيف";
            StringBuilder sb = new StringBuilder();
            sb.append(verdict == PolicyVerdict.BLOCKED ? "🔴 مرفوض" : "🟡 تحذير");
            sb.append(" — ");
            violations.forEach(v -> sb.append(v.rule).append(' '));
            sb.append("(").append(violations.size()).append(" انتهاك)");
            return sb.toString().trim();
        }

        /** تفسير بالعربي لكل انتهاك */
        public String explain() {
            if (violations.isEmpty()) return "الكود نظيف — لا انتهاكات.";
            StringBuilder sb = new StringBuilder();
            sb.append("تقرير PolicyEngine:\n");
            sb.append("━━━━━━━━━━━━━━━━━━━━━━━━━\n");
            sb.append(String.format("الحكم    : %s\n", verdict));
            sb.append(String.format("مستوى الخطر: %d/100\n", riskScore));
            sb.append(String.format("الثقة العامة: %.0f%%\n\n", overallConfidence * 100));
            for (PolicyViolation v : violations) {
                sb.append(String.format("• %s — %s\n", v.rule, v.description));
                sb.append(String.format("  السطر %d: %s\n", v.line, v.evidence));
                sb.append(String.format("  الخطورة: %s  |  الثقة: %.0f%%\n",
                    v.critical ? "حرج" : "تحذير", v.confidence * 100));
                sb.append("\n");
            }
            return sb.toString().trim();
        }
    }

    // ═══════════════════════════════════════════════════
    // Rule Definition
    // ═══════════════════════════════════════════════════

    private static class Rule {
        final String  id;
        final String  description;
        final Pattern pattern;
        final boolean critical;
        final double  baseConfidence; // ثقة أساسية للـ rule

        Rule(String id, String description, String regex,
             boolean critical, double baseConfidence) {
            this.id             = id;
            this.description    = description;
            this.pattern        = Pattern.compile(regex,
                Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
            this.critical       = critical;
            this.baseConfidence = baseConfidence;
        }
    }

    // ── Whitelist: مصطلحات مسموحة تُقلّل الـ confidence ──
    private static final Set<String> WHITELIST_TERMS = new HashSet<>(Arrays.asList(
        "test_keylog", "mock_encrypt", "example_shell", "demo_brute"
    ));

    /** أضف مصطلح للـ whitelist (مثلاً اسم test function) */
    public static void addWhitelist(String term) {
        WHITELIST_TERMS.add(term.toLowerCase());
    }

    /** تحقق لو الـ evidence يحتوي whitelist term */
    public static boolean isWhitelisted(String evidence) {
        String ev = evidence.toLowerCase();
        return WHITELIST_TERMS.stream().anyMatch(ev::contains);
    }

    /** أضف rule مخصصة في runtime */
    public static void addRule(String id, String description,
                                String regex, boolean critical,
                                double baseConfidence) {
        RULES.add(new Rule(id, description, regex, critical, baseConfidence));
    }

    // ═══════════════════════════════════════════════════
    // Rules — compiled في static init (مرة واحدة)
    // ═══════════════════════════════════════════════════

    private static final List<Rule> RULES;

    static {
        RULES = new ArrayList<>();

        // ── Critical rules (BLOCKED) ──
        RULES.add(new Rule("POL001", "حذف ملفات النظام",
            "(shutil\\.rmtree|os\\.remove)\\s*\\(.*?(/etc|/sys|/bin|/usr)",
            true, 0.95));

        RULES.add(new Rule("POL002", "سرقة بيانات المستخدم",
            "(keylog|keystroke|capture_password|hook_keyboard)",
            true, 0.90));

        RULES.add(new Rule("POL003", "ransomware",
            "(encrypt|AES|Fernet)[\\s\\S]{0,200}(walk|glob)[\\s\\S]{0,100}\\.(doc|pdf|xls|jpg)",
            true, 0.85));

        RULES.add(new Rule("POL004", "أدوات اختراق",
            "(brute[_\\-\\s]?force|reverse[_\\-\\s]?shell|bind[_\\-\\s]?shell|meterpreter)",
            true, 0.95));

        RULES.add(new Rule("POL005", "تنفيذ كود مشفر",
            "(exec|eval)\\s*\\(\\s*(base64|b64decode|codecs\\.decode)",
            true, 0.90));

        RULES.add(new Rule("POL006", "تجاوز صلاحيات النظام",
            "(setuid\\s*\\(0\\)|chmod\\s+[0-7]*7[0-7][0-7]|sudo\\s+.*--no-ask-passwd)",
            true, 0.88));

        // ── Warning rules (WARNING) ──
        RULES.add(new Rule("POL007", "تحميل وتشغيل ملفات خارجية",
            "(urlretrieve|requests\\.get)[\\s\\S]{0,100}\\.(exe|sh|bat|ps1)[\\s\\S]{0,100}(run|subprocess|exec)",
            false, 0.75));

        RULES.add(new Rule("POL008", "إرسال بيانات حساسة",
            "requests\\.(post|put)\\s*\\([\\s\\S]{0,200}(password|api_key|secret|token)",
            false, 0.70));

        RULES.add(new Rule("POL009", "تعديل ملفات النظام",
            "open\\s*\\(\\s*['\"]\\s*(/etc/hosts|/etc/crontab|/etc/sudoers|/etc/passwd)",
            false, 0.85));

        RULES.add(new Rule("POL010", "أوامر تدمير البيانات",
            "(dd\\s+if=|mkfs\\s*\\.|:\\s*\\(\\s*\\)\\s*\\{|>\\s*/dev/sd[a-z])",
            false, 0.92));

        RULES.add(new Rule("POL011", "كود obfuscation مشبوه",
            "([A-Za-z0-9+/]{80,}={0,2}|\\\\x[0-9a-fA-F]{2}(\\\\x[0-9a-fA-F]{2}){15,})",
            false, 0.65));
    }

    // ═══════════════════════════════════════════════════
    // Preprocessor — يحذف comments و string literals
    // ═══════════════════════════════════════════════════

    /**
     * يُعيد الكود مع استبدال:
     * - Python/JS single-line comments بـ placeholder
     * - string literals بـ placeholder (يحتفظ بالطول)
     * لتجنب false positives من `# keylog` أو `"password"` في الـ docs
     */
    static String stripCommentsAndStrings(String code) {
        // ١. single-line comments: # ... أو // ...
        code = code.replaceAll("(?m)(#|//).*$", "/* COMMENT */");

        // ٢. multi-line comments: /* ... */
        code = code.replaceAll("/\\*[\\s\\S]*?\\*/", "/* COMMENT */");

        // ٣. triple-quoted strings (Python)
        code = code.replaceAll("\"\"\"[\\s\\S]*?\"\"\"|'''[\\s\\S]*?'''", "\"STRIPPED\"");

        // ٤. double-quoted strings (بسيطة — بدون escape معقد)
        code = code.replaceAll("\"(?:[^\"\\\\]|\\\\.)*\"", "\"STRIPPED\"");

        // ٥. single-quoted strings
        code = code.replaceAll("'(?:[^'\\\\]|\\\\.)*'", "'STRIPPED'");

        return code;
    }

    // ═══════════════════════════════════════════════════
    // Evidence Extractor — يُعيد السطر كاملاً مع context
    // ═══════════════════════════════════════════════════

    /**
     * بدل قطع الـ match بـ 80 حرف ثابت،
     * يُعيد السطر الكامل الذي يحتوي الـ match
     * مع رقم السطر
     */
    private static int[] findLineAndEvidence(String original, int matchStart) {
        // أوجد رقم السطر
        int line = 1;
        for (int i = 0; i < matchStart && i < original.length(); i++) {
            if (original.charAt(i) == '\n') line++;
        }
        return new int[]{ line };
    }

    private static String extractLine(String code, int matchStart) {
        int lineStart = matchStart;
        while (lineStart > 0 && code.charAt(lineStart - 1) != '\n') lineStart--;
        int lineEnd = matchStart;
        while (lineEnd < code.length() && code.charAt(lineEnd) != '\n') lineEnd++;
        String line = code.substring(lineStart, lineEnd).trim();
        // قصّ لو طويل جداً
        return line.length() > 120 ? line.substring(0, 117) + "..." : line;
    }

    // ═══════════════════════════════════════════════════
    // Confidence Calculator
    // ═══════════════════════════════════════════════════

    /**
     * يحسب confidence لانتهاك واحد بناءً على:
     * - baseConfidence من الـ Rule
     * - عدد matches في الكود (+bonus)
     * - وجود whitelist term (-penalty)
     * - طول الـ evidence (قصير = أدق)
     */
    public static double confidence(Rule rule, String code,
                                     String evidence, int matchCount) {
        double conf = rule.baseConfidence;

        // bonus: كلما تكرر النمط زاد اليقين
        if (matchCount > 1) conf = Math.min(1.0, conf + 0.05 * (matchCount - 1));

        // penalty: لو الـ evidence يحتوي whitelist term
        if (isWhitelisted(evidence)) conf *= 0.40;

        // penalty: لو الـ evidence قصير جداً (< 10 chars) — match غير دقيق
        if (evidence.trim().length() < 10) conf *= 0.70;

        return Math.round(conf * 100.0) / 100.0; // round to 2 decimals
    }

    // ═══════════════════════════════════════════════════
    // check() — Entry Point
    // ═══════════════════════════════════════════════════

    public static PolicyResult check(String code) {
        if (code == null || code.isBlank()) {
            return new PolicyResult(PolicyVerdict.ALLOWED, Collections.emptyList(), 0);
        }

        String cleanCode = stripCommentsAndStrings(code);
        List<PolicyViolation> violations = new ArrayList<>();

        for (Rule rule : RULES) {
            Matcher m = rule.pattern.matcher(cleanCode);
            if (!m.find()) continue;

            int matchStart = m.start();

            // عدّ كل الـ matches لحساب الـ confidence
            int matchCount = 1;
            while (m.find()) matchCount++;

            int    line     = findLineAndEvidence(cleanCode, matchStart)[0];
            String evidence = extractLine(code, matchStart);
            double conf     = confidence(rule, cleanCode, evidence, matchCount);

            violations.add(new PolicyViolation(
                rule.id, rule.description, evidence, line, rule.critical, conf
            ));
        }

        int riskScore = 0;
        for (PolicyViolation v : violations)
            riskScore += v.critical ? (int)(40 * v.confidence) : (int)(20 * v.confidence);
        riskScore = Math.min(100, riskScore);

        boolean hasCritical = violations.stream()
            .anyMatch(v -> v.critical && v.confidence >= 0.5);
        PolicyVerdict verdict = hasCritical           ? PolicyVerdict.BLOCKED
                              : !violations.isEmpty() ? PolicyVerdict.WARNING
                              : PolicyVerdict.ALLOWED;

        return new PolicyResult(verdict, violations, riskScore);
    }

    // ═══════════════════════════════════════════════════
    // checkAll() — فحص مجموعة ملفات
    // ═══════════════════════════════════════════════════

    /**
     * يفحص map من fileName → code
     * يُعيد map من fileName → PolicyResult
     */
    public static Map<String, PolicyResult> checkAll(Map<String, String> files) {
        Map<String, PolicyResult> results = new LinkedHashMap<>();
        if (files == null) return results;
        for (Map.Entry<String, String> entry : files.entrySet()) {
            results.put(entry.getKey(), check(entry.getValue()));
        }
        return results;
    }

    /**
     * يُعيد true لو أي ملف في الـ map محظور
     */
    public static boolean anyBlocked(Map<String, PolicyResult> results) {
        return results.values().stream()
            .anyMatch(r -> r.verdict == PolicyVerdict.BLOCKED);
    }

    /**
     * يُعيد ملخص نصي لنتائج checkAll
     */
    public static String summarizeAll(Map<String, PolicyResult> results) {
        StringBuilder sb = new StringBuilder();
        sb.append("═══ PolicyEngine Batch Report ═══\n");
        int blocked = 0, warned = 0, clean = 0;
        for (Map.Entry<String, PolicyResult> e : results.entrySet()) {
            PolicyResult r = e.getValue();
            sb.append(String.format("%-30s → %s\n", e.getKey(), r.summary));
            if (r.verdict == PolicyVerdict.BLOCKED)  blocked++;
            else if (r.verdict == PolicyVerdict.WARNING) warned++;
            else clean++;
        }
        sb.append(String.format("\nملخص: 🔴 %d محظور  🟡 %d تحذير  ✅ %d نظيف",
            blocked, warned, clean));
        return sb.toString();
    }
}
